src/source/src/table.rs

```diff
@@ -146,16 +207,40 @@ impl TableDmlHandle {
 #[derive(Debug)]
 pub struct TableStreamReader {
     /// The receiver of the changes channel.
-    rx: mpsc::Receiver<(StreamChunk, oneshot::Sender<usize>)>,
+    rx: mpsc::UnboundedReceiver<(TxnMsg, oneshot::Sender<usize>)>,
 }
 
 impl TableStreamReader {
     #[try_stream(boxed, ok = StreamChunkWithState, error = RwError)]
+    pub async fn into_stream_for_source_reader_test(mut self) {
+        while let Some((txn_msg, notifier)) = self.rx.recv().await {
+            // Notify about that we've taken the chunk.
+            match txn_msg {
+                TxnMsg::Begin(_) | TxnMsg::End(_) | TxnMsg::Rollback(_) => {
+                    _ = notifier.send(0);
+                }
+                TxnMsg::Data(_, chunk) => {
+                    _ = notifier.send(chunk.cardinality());
+                    yield chunk.into();
+                }
+            }
+        }
+    }
+
+    #[try_stream(boxed, ok = TxnMsg, error = RwError)]
     pub async fn into_stream(mut self) {
-        while let Some((chunk, notifier)) = self.rx.recv().await {
+        while let Some((txn_msg, notifier)) = self.rx.recv().await {
             // Notify about that we've taken the chunk.
-            _ = notifier.send(chunk.cardinality());
-            yield chunk.into();
+            match &txn_msg {
+                TxnMsg::Begin(_) | TxnMsg::End(_) | TxnMsg::Rollback(_) => {
+                    _ = notifier.send(0);
```

Comment:
IIUC, we still don't provide durability as we immediately notify the sender here.